File: android/src/main/kotlin/io/sentient/android/settings/personalities/PersonalitiesViewModel.kt
// ---------------------------------------------------------------------------
// PersonalitiesViewModel — Personalities settings page. Unlike the draft+Save pages,
// each mutation (create / update-body / activate / delete) is its OWN restart-on-write
// FSM run; the list is refetched after every Ready (webui: trust post-restart truth).
// The VM holds only the server list + FSM progress; per-card expand / edit-draft /
// new-form state is view-local in the screen.
// ---------------------------------------------------------------------------
package io.sentient.android.settings.personalities

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import io.sentient.mobiledata.di.SettingsComponent
import io.sentient.mobiledata.result.SentientResult
import io.sentient.mobiledata.usecase.settings.ApplyState
import io.sentient.mobiledata.usecase.settings.ProfileMutation
import io.sentient.mobilesdk.log.createLogger
import io.sentient.mobilesdk.settings.Personality
import io.sentient.mobilesdk.settings.PersonalityList
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

/** Personalities page state. `list` is server truth; progress fields fold the active FSM run. */
data class PersonalitiesUiState(
    val loading: Boolean = true,
    val loadError: String? = null,
    val list: PersonalityList? = null,
    val saving: Boolean = false,
    val restarting: Boolean = false,
    val alreadyApplying: Boolean = false,
    val applyError: String? = null,
) {
    val busy: Boolean get() = saving || restarting
    val personalities: List<Personality> get() = list?.personalities ?: emptyList()
    val activeName: String? get() = list?.activeName
}

class PersonalitiesViewModel(private val component: SettingsComponent) : ViewModel() {
    private val log = createLogger("android", "settings", "personalities-vm")

    private val _ui = MutableStateFlow(PersonalitiesUiState())
    val ui: StateFlow<PersonalitiesUiState> = _ui.asStateFlow()

    init {
        reload()
    }

    fun create(name: String, body: String) =
        runMutation(ProfileMutation.CreatePersonality(name.trim(), body), "create")

    fun updateBody(name: String, body: String) =
        runMutation(ProfileMutation.UpdatePersonality(name, body), "update")

    fun activate(name: String) = runMutation(ProfileMutation.ActivatePersonality(name), "activate")

    fun delete(name: String) = runMutation(ProfileMutation.DeletePersonality(name), "delete")

    private fun runMutation(mutation: ProfileMutation, op: String) {
        if (_ui.value.busy) return
        log.info("mutation", mapOf("op" to op))
        viewModelScope.launch {
            component.applyProfileChange(mutation).collect { st ->
                _ui.update { it.foldApply(st) }
                if (st is ApplyState.Ready) reload()
            }
        }
    }

    private fun reload() {
        viewModelScope.launch {
            when (val r = component.profileRepository.listPersonalities()) {
                is SentientResult.Success -> _ui.update { it.copy(loading = false, list = r.data, loadError = null) }
                is SentientResult.Failure -> {
                    log.warn("reload.failed", mapOf("kind" to r.error.kind))
                    _ui.update { it.copy(loading = false, loadError = r.error.userMessage) }
                }
                is SentientResult.Loading -> Unit
            }
        }
    }
}

/** Fold one FSM transition into flat progress fields. */
internal fun PersonalitiesUiState.foldApply(state: ApplyState): PersonalitiesUiState = when (state) {
    ApplyState.Saving -> copy(saving = true, restarting = false, alreadyApplying = false, applyError = null)
    ApplyState.Restarting -> copy(saving = false, restarting = true)
    is ApplyState.Ready -> copy(saving = false, restarting = false, alreadyApplying = false, applyError = null)
    ApplyState.AlreadyApplying -> copy(saving = false, restarting = false, alreadyApplying = true)
    is ApplyState.Failed -> copy(saving = false, restarting = false, applyError = state.error.userMessage)
    ApplyState.Idle -> this
}

File: ios/App/Settings/Personalities/PersonalitiesViewModel.swift
// ---------------------------------------------------------------------------
// PersonalitiesViewModel — the Personalities page state holder. Unlike the
// draft+Save pages, every action here is IMPERATIVE: Activate / Delete / Create
// each run their own ApplyProfileChangeUseCase FSM (all restart-on-write), and
// the list is refetched on Ready to trust post-restart server truth.
//
// A single `op` projection tracks the one in-flight mutation (saving → restarting
// → ready | alreadyApplying | failed), surfaced as a page banner. Personalities'
// names are immutable after create (the create form is the only place a name is
// set), so there is no page-level dirty draft.
// ---------------------------------------------------------------------------
import Foundation
import MobileData

@MainActor
@Observable
final class PersonalitiesViewModel {
    enum Phase: Equatable {
        case loading
        case ready
        case failed(String)
    }

    enum Op: Equatable {
        case idle
        case saving
        case restarting
        case alreadyApplying
        case applied
        case failed(String)
    }

    private(set) var phase: Phase = .loading
    private(set) var op: Op = .idle
    private(set) var personalities: [Personality] = []
    private(set) var activeName: String?

    private let settings: SettingsComponent
    private let log = AppLog("settings", "personalities-vm")

    init(settings: SettingsComponent) {
        self.settings = settings
    }

    var isBusy: Bool { op == .saving || op == .restarting }
    var operationError: String? {
        if case .failed(let message) = op { return message }
        if case .alreadyApplying = op { return "Another change is already applying. Try again in a moment." }
        return nil
    }

    func load() async {
        log.info("load")
        do {
            let result = try await settings.profileRepository.listPersonalities()
            switch onEnum(of: result) {
            case .success(let s):
                personalities = s.data.personalities
                activeName = s.data.activeName
                phase = .ready
                log.info("load.ready count=\(personalities.count)")
            case .failure(let f):
                phase = .failed(f.error.userMessage)
                log.warn("load.failed kind=\(f.error.kind)")
            case .loading:
                phase = .loading
            }
        } catch is CancellationError {
        } catch {
            phase = .failed("Couldn't load personalities.")
            log.warn("load.threw")
        }
    }

    func activate(_ name: String) async {
        log.info("activate nameLen=\(name.count)")
        await run(ProfileMutationActivatePersonality(name: name), label: "activate")
    }

    func delete(_ name: String) async {
        log.info("delete nameLen=\(name.count)")
        await run(ProfileMutationDeletePersonality(name: name), label: "delete")
    }

    /// Create a personality. Returns true on success so the caller can dismiss the
    /// create form. A blank name is rejected client-side (no request fired).
    func create(name: String, body: String) async -> Bool {
        let trimmed = name.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            op = .failed("Name can't be empty.")
            return false
        }
        log.info("create nameLen=\(trimmed.count)")
        await run(ProfileMutationCreatePersonality(name: trimmed, body: body), label: "create")
        return op == .applied
    }

    /// Fold one mutation's FSM; refetch the list on Ready. Never throws across the
    /// boundary — a failure sets `op = .failed(...)` and preserves the list.
    private func run(_ mutation: any ProfileMutation, label: String) async {
        for await state in settings.applyProfileChange.invoke(mutation: mutation) {
            switch onEnum(of: state) {
            case .idle: break
            case .saving: op = .saving
            case .restarting: op = .restarting
            case .ready:
                log.info("\(label).ready")
                await load()
                op = .applied
            case .alreadyApplying:
                op = .alreadyApplying
                log.warn("\(label).already-applying")
            case .failed(let f):
                op = .failed(f.error.userMessage)
                log.warn("\(label).failed")
            }
        }
    }
}
